Add new navbar with JS logic

YashRE42 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
This re-stylizes the top navbar and adds JS logic to keep it functional.
A lot of detail on what changes are still needed as part of this are in [this discussion on czo](https://chat.zulip.org/#narrow/stream/6-frontend/topic/new.20nav-bar.20with.20sub.20count.20and.20description).

**Testing Plan:** <!-- How have you tested? -->
The unit tests for this are yet to be added.

**GIFs or Screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->
Will add screenshots and GIFs soon.

<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

YashRE42 commented:
The commit order between this and my branch is being changed in the middle of the PR (reason why CI is failing). I'm not sure why that is happening but I'm working on finding a solution.
The correct order for commits is:
1) navbar: Refactor to nested selectors in scss. ea8775b
2) navbar: Styling for logged-in top navbar. 54e4bdf
3) navbar: Reuse search icon for search open/close. 2bac3a0
4) navbar: Remove topic breadcrumb. 8887962
5) navbar: Seperate private from pm-with. 33b819f
6) navbar: Add JS logic for displaying elements. db0d2ed
7) navbar: Color lock/hash to stream color. 38f27d7

rishig commented:
Thanks @YashRE42! I haven't looked at the code, but first round of product comments at https://paper.dropbox.com/doc/Nav-bar-roadmap--AZQp2zrnWfybr0Kh2tDAUFqOAg-xNIWORJrq1M9ZT9xht02O.

YashRE42 commented:
@timabbott this is still in-progress and I'm going to make all the changes I've spoken with rishi about, but I'd like to know if the 6th commit: [`navbar: Add JS logic for displaying elements. ab7ffaa`](https://github.com/zulip/zulip/pull/11894/commits/ab7ffaa8b310695b078a852e4361a9c8d3edad5e) is an acceptable approach, I'd appreciate if you could have a look at it.

timabbott commented on static/js/tab_bar.js:
```diff
@@ -138,6 +138,103 @@ exports.colorize_tab_bar = function () {
+        $(".navbar-lock").show();
+    } else {
+        $(".hash").show();
+    }
```
I would definitely do this function with a handlebars template that we render and then swap in as the new HTML for that area.  As a style/readability rule, we don't construct complex HTML inside JS code in this project.

timabbott commented:
The high-level approach is reasonable; see my comment on the details of how to generate HTML in Zulip.

YashRE42 commented on static/templates/tab_bar.handlebars:
```diff
@@ -1,21 +1,17 @@
-    <li class="{{active}} {{cls}}" {{#if data}}data-name="{{data}}"{{/if}}>
-        {{#if icon}}
-        <i class="fa fa-angle-right" aria-hidden="true"></i>
-        {{/if}}
```
From what I understand from the git history, we stopped sending "icon" to this template 6 years ago in commit `a6486a1fa`, and yet this was never removed from the template itself. This is partially why I moved towards such a refactor.

YashRE42 commented:
@timabbott I made significant changes in the last commit. ([navbar: Refactor tab_bar. 90b059c](https://github.com/zulip/zulip/pull/11894/commits/90b059cc78745fc4160390f279e9ef45403af2a0))
I think this really cleans up tab_bar.js and tab_bar.handlebar. Please have a look.

timabbott commented:
A few random notes from testing this:
* The `cursor` CSS is wrong for the stream description text and subscriber count -- the mouse behaves like it would over a text box.
* We should hover over the subscriber count explaining that it means "6 users are subscribed to #streamname" or something.
* When you click the magnifying glass icon, it should put focus in the text box after transitioning.
* There's a weird white line in night mode at the edges of the top navbar:
![image](https://user-images.githubusercontent.com/2746074/54564866-22393600-498a-11e9-9345-a6310a5a249f.png)

YashRE42 commented:
- ~The cursor CSS is wrong for the stream description text and subscriber count -- the mouse behaves like it would over a text box.~
- ~We should hover over the subscriber count explaining that it means "6 users are subscribed to #streamname" or something.~

![cursor and tooltip](https://user-images.githubusercontent.com/33805964/54596699-3436d980-4a5b-11e9-9764-10d2b43aab26.png)

- ~When you click the magnifying glass icon, it should put focus in the text box after transitioning.~
- ~open search-bar by hotkeys~

[webm](https://chat.zulip.org/user_uploads/2/3e/RgOmlU6VtY-OLzc1ayRIELem/Search-icon-and-k-key.webm)

- ~There's a weird white line in night mode at the edges of the top navbar:~

![nightbar-search-closed](https://user-images.githubusercontent.com/33805964/54598148-f25b6280-4a5d-11e9-8fd7-c57ed6dff17f.png)
![night-bar-search-open](https://user-images.githubusercontent.com/33805964/54598143-eff90880-4a5d-11e9-9cbc-ea0bf3761f70.png)

I'd like opinions on the following:
1. Clicking the search icon should move the cursor to the end of the search and not highlight the existing filter text

for:
  - When a user enters a stream and clicks search, there's better chances that they want to do a search    in the stream and this should be available at one click.

 against:
 - It's easy to go from "filter selected" to the end of the search query by pressing the right arrow key.
- Searching by hotkey '/' and 'cntrl-K' select the text and we want to be consistent with existing behavior.


2. When search is locked open, ie the search-icon is disabled, we should make it trigger a search on click, but this would probably take a reasonable amount of time to get right and it's worth thinking if we want it.

for:
- Seems more intuitive than just an icon.

Against:
- This would be inconsistent with what the icon otherwise does.

Another issue that we'd want to fix, but would take a reasonable amount of time to get right would be:

3. Only the icon should be clickable or the entire square around it (compacted search bar) should be clickable and open the search bar. It shouldn't be possible to click the compacted search bar if you accidentally miss the icon.

![missed icon](https://user-images.githubusercontent.com/33805964/54600209-410afb80-4a62-11e9-90ce-8c0c24d879b0.png)


I'd like to know if everything else here is okay before trying to crack any of these last three, and if we'd want to have those changes now rather than as follow ups or after testing (and node tests).

YashRE42 commented:
- ~The cursor CSS is wrong for the stream description text and subscriber count -- the mouse behaves like it would over a text box.~
- ~We should hover over the subscriber count explaining that it means "6 users are subscribed to #streamname" or something.~

![cursor and tooltip](https://user-images.githubusercontent.com/33805964/54596699-3436d980-4a5b-11e9-9764-10d2b43aab26.png)

- ~When you click the magnifying glass icon, it should put focus in the text box after transitioning.~
- ~open search-bar by hotkeys~

[webm](https://chat.zulip.org/user_uploads/2/3e/RgOmlU6VtY-OLzc1ayRIELem/Search-icon-and-k-key.webm)

- ~There's a weird white line in night mode at the edges of the top navbar:~

![nightbar-search-closed](https://user-images.githubusercontent.com/33805964/54598148-f25b6280-4a5d-11e9-8fd7-c57ed6dff17f.png)
![night-bar-search-open](https://user-images.githubusercontent.com/33805964/54598143-eff90880-4a5d-11e9-9cbc-ea0bf3761f70.png)

I'd like opinions on the following:
1. Clicking the search icon should move the cursor to the end of the search and not highlight the existing filter text

for:
  - When a user enters a stream and clicks search, there's better chances that they want to do a search    in the stream and this should be available at one click.

 against:
 - It's easy to go from "filter selected" to the end of the search query by pressing the right arrow key.
- Searching by hotkey '/' and 'cntrl-K' select the text and we want to be consistent with existing behavior.


2. When search is locked open, ie the search-icon is disabled, we should make it trigger a search on click, but this would probably take a reasonable amount of time to get right and it's worth thinking if we want it.

for:
- Seems more intuitive than just an icon.

Against:
- This would be inconsistent with what the icon otherwise does.

Another issue that we'd want to fix, but would take a reasonable amount of time to get right would be:

3. Only the icon should be clickable or the entire square around it (compacted search bar) should be clickable and open the search bar. It shouldn't be possible to click the compacted search bar if you accidentally miss the icon.

![missed icon](https://user-images.githubusercontent.com/33805964/54600209-410afb80-4a62-11e9-90ce-8c0c24d879b0.png)


I'd like to know if everything else here is okay before trying to crack any of these last three, and if we'd want to have those changes now rather than as follow ups or after testing (and node tests).
(Also, if I'd like to know if I should squash everything down to ~ two commits or wait.)

timabbott commented:
> Clicking the search icon should move the cursor to the end of the search and not highlight the existing filter text

I think we want the highlight to be there, because that's what makes it easy to delete the existing filter if one wants to.  I think hitting `right arrow` should let the user further refine the search, and that's not too annoying.

> When search is locked open, ie the search-icon is disabled, we should make it trigger a search on click, but this would probably take a reasonable amount of time to get right and it's worth thinking if we want it.

Yeah, I don't think we need to do anything here.

In terms of next steps, I'd love it if you cleaned up the commit history to be super readable, and then I can push on trying to get this test-deployed on chat.zulip.org for more feedback (and possibly merge parts that are mergable).

YashRE42 commented:
@timabbott I've squashed this down and hopefully the commit message gives a sufficient overview. There are still some kinks, related to responsiveness and the width of the navbar for long stream descriptions.

YashRE42 commented:
@timabbott This last commit that I've pushed is one way of giving priority to the width of the stream name, it's not an optimal solution. It doesn't handle extremely long stream names, which is a case we'd want to add in. We'd also want something in `static/js/resize.js` that would handle this, rather than only triggering whenever a stream is selected. However, I think that this is okay for test deploy, because I'm not aware of any other glaring issues.

YashRE42 commented:
@timabbott Is there anything more I can do right now to help take this forward?

YashRE42 commented:
@rishig Can you try this out and see if it looks like what we'd want in the end?

rishig commented:
Thanks @YashRE42, looks great! A few comments:
* The on-hover mouse icon suggests that elements are clickable, when they aren't
* A small animation that slides the magnifying glass from right to left when you click on it might make it easier to understand what's going on.
* No live update of stream description
* There are some styling tweaks we'll need to make, but that can be done in post. 
A small one is that there is too much space to the right of the `#`, and the `#` should probably be a little bolder

Followups for after we test this out:
* The lock icon to the left of stream names on recipient bars doesn't have enough right-padding (the new nav bar makes this look worse)
![image](https://user-images.githubusercontent.com/890911/55639659-c2f76600-577e-11e9-8a4e-3b21346345c7.png)

* The topic is now not prominent enough in topic narrows

* The new nav bar adds a bit to our visual hierarchy problem. Possibly the top area should have a slight background highlight, or the separating line should be a bit bolder (or have a drop shadow)?

YashRE42 commented:
@rishig I've made the changes you've asked, and also there's some other things to consider.
Please let me know if there's something I should do on this right now or if it's good for now.

> No live update of stream description [or stream name]

Done

> The on-hover mouse icon suggests that elements are clickable, when they aren't

Done

>  A small animation that slides the magnifying glass from right to left when you click on it might make it easier to understand what's going on.

This is something that we mentioned in the [paper](https://paper.dropbox.com/doc/Nav-bar-roadmap--AasMHI~Ihz7Cijc1__~wMavcAg-xNIWORJrq1M9ZT9xht02O#:uid=314461265353820905503559&h2=Where-we%E2%80%99d-want-to-go-after-we) it'll probably be hard to quickly get a consensus on the animation itself, so maybe we'd want to shift this to something we handle after a test deploy.

> There are some styling tweaks we'll need to make, but that can be done in post.
>  A small one is that there is too much space to the right of the `#`, and the `#` should probably be a little bolder

We were making a color space call to lighten it by 0.2, which I removed, but this is probably not enough and we'd need to handle this later.
![comparativeOfColor](https://user-images.githubusercontent.com/33805964/55668530-2a014180-5889-11e9-9047-894b04e1f495.png)

rishig commented:
Thanks @YashRE42! 
* Clicking on search should set the focus on the end of the text box, and add a space.

Other followups: 
* Clearing the search also clears the narrow. E.g. click `Private messages`, then click the search icon, then click the `X` that replaces the search icon.
This is the current behavior in master, but in master there is less reason to ever click the `X`. 
* There should be a little more space to the left of the mag here:
![image](https://user-images.githubusercontent.com/890911/55748049-85027800-59f2-11e9-957c-dd7cc56fd6f4.png)

I guess we can copy the followups into the doc?

YashRE42 commented:
> > Clicking the search icon should move the cursor to the end of the search and not highlight the existing filter text
> 
> I think we want the highlight to be there, because that's what makes it easy to delete the existing filter if one wants to. I think hitting `right arrow` should let the user further refine the search, and that's not too annoying.

So I had fixed this up and at some point regressed on it, but I've fixed it up again, just needed to reorder the call to `.select()` to be after the navbar is expanded.
[recording for current behavior.](https://chat.zulip.org/user_uploads/2/e1/bpkyLRUs1vBdK6jNc7X9j3Xd/Screencast-from-Tuesday-09-April-2019-050550-IST.webm)
>  Clicking on search should set the focus on the end of the text box, and add a space.

By "clicking on search" I assume you mean clicking on the search icon?
if so:
* I can do this, but I foresee people being upset about the design change here, especially if the change to the "X" isn't done at the same time because then clearing the filter is less easy.

if not:
* I think [this behavior (recording)](https://chat.zulip.org/user_uploads/2/3d/iDBpvw3leIit5wKGx5JcM9Sj/Screencast-from-Tuesday-09-April-2019-045016-IST.webm) is very good and it would be complicated to change.



> Clearing the search also clears the narrow. E.g. click `Private messages`, then click the search icon, then click the `X` that replaces the search icon.
>       This is the current behavior in master, but in master there is less reason to ever click the `X`.

rishig commented:
hm, the recording links aren't working for me.

YashRE42 commented:
@rishig replaced the recordings with gifs,
Please let me know if there's any changes that I'd need to do right now.
Also, should revive the CZO discussion for this in the hopes that it'd make progress faster?

rishig commented:
cool, thanks! test deploy sounds good.

timabbott commented:
I'd like to work intensely towards merging this.  As part of doing that, we'll need to rework the commit structure.  I stored the last version of this branch with the original commit structure as `timabbott/searchbox-original-branch`, squashed everything, and have started on extracting clean preparatory commits that we can merge now.  Here's a partial TODO list:
* [x] Extract the changes to `.search_icon`; that's independent and should be mergable early.
* [ ] Figure out how to convert the changes to bootstrap-typeahead into something maintainable.  We have a policy of not including any non-generic code in our patches to the typeahead library.  @aero31aero do you have any ideas for how to do this?
* [ ] I'd like to understand the explanation behind why this changeset involves introducing conditionals based on the `left_side_userlist` state; @YashRE42 can you explain what's going on there?  I would not have expected that changes to this component would require so much new code in `media.scss` or passing that version of the `left_side_userlist` flag into the template rendering.  
* [ ] The definition of `.search_closed:hover` should use SCSS nesting with `&:hover`.
* [ ] Is it intentional that the CSS changes for `.searchbox` and `.searchbox_legacy` are different, or is that an oversight?  
* [ ] Figure out a less hacky way to implement the goals of `set_max_width_of_descriptions`.  Could we do this using CSS overflow properties?

aero31aero commented:
> Figure out how to convert the changes to bootstrap-typeahead into something maintainable

The following patch could work, but maybe there's a better option that I can't see because I don't completely understand the intent behind the code yet. Maybe we should first split the commits.

```diff
diff --git static/js/zcommand.js static/js/zcommand.js
index 193085daa..edef1070c 100644
--- static/js/zcommand.js
+++ static/js/zcommand.js
@@ -16,6 +16,10 @@ What in the heck is a zcommand?
 
 */
 
+$(document).on('zuliptypeaheadclosed', (e) => {
+    console.log("We are about to close a typeahead", e);
+})
+
 exports.send = function (opts) {
     const command = opts.command;
     const on_success = opts.on_success;
diff --git static/third/bootstrap-typeahead/typeahead.js static/third/bootstrap-typeahead/typeahead.js
index 1d92ddf4d..f62daa168 100644
--- static/third/bootstrap-typeahead/typeahead.js
+++ static/third/bootstrap-typeahead/typeahead.js
@@ -49,6 +49,15 @@
  *   Our custom changes include all mentions of this.header, some CSS changes
  *   in compose.scss and splitting $container out of $menu so we can insert
  *   additional HTML before $menu.
+ *
+ * 4. Hook for dismissing typeahead:
+ *
+ *   We emit an event on dismissing a typeahead, which can be used by other
+ *   parts of Zulip code. This change is mainly used by the JS based navbar.
+ *   Look at https://github.com/zulip/zulip/pull/11894#issuecomment-577469170.
+ *
+ *   Our custom change includes emitting an event inside the .hide() method of
+ *   the typeahead.
  * ============================================================ */
 
 !function($){
@@ -161,6 +170,7 @@
     }
 
   , hide: function () {
+      $(document).trigger("zuliptypeaheadclosed"); // Zulip specific change.
       this.$container.hide()
       this.shown = false
       return this

```

YashRE42 commented:
I wanted to update that I have been looking at this, the few rebases that have happened here might have caused a few breakages, so I'm going to try and rebase the oldest local version I had with master and hopefully that'll make it easier to extract prep commits. (I noticed a regression with the positioning of the search icon, which was probably caused by rebase)

timabbott commented:
OK.  Keep us updated!

YashRE42 commented:
Anders moved the changes on this branch when media.scss was deleted, so I didn't need to do much, just dropped media.scss.
I tried to move the changes to subs.js as a prep commit, `$(".narrow_description")` would return `undefined`.
The regression I thought I saw only happens if you don't reload after changing `left_side_userlist` and are already in a narrow view, and it goes away if you refresh, I don't think it's important.

We only use `left_side_userlist` in the template to add a class which some styles hit to adjust margins and positions.

timabbott commented on static/js/subs.js:
```diff
@@ -164,6 +167,13 @@ exports.update_stream_description = function (sub, description, rendered_descrip
+    const stream_description = $(".narrow_description");
+    if (stream_description !== undefined) {
+        stream_description.html("");
+        stream_description.append(sub.rendered_description);
```
This commit could call into a similar no-op function in `tab_bar.js`.  

Also, would it be better to do `stream_description.html(sub.rendered_description)` rather than clearing and then re-adding?

YashRE42 commented:
I had misunderstood what you meant by the call point being wrong. I have now moved the call from `stream_edit.js` to `sub.js`.

YashRE42 commented:
For `set_max_width_of_descriptions`... I think we could just set constant values instead of performing complex calculations and it would look fine in most cases.

YashRE42 commented on static/styles/night_mode.scss:
```diff
@@ -218,8 +216,27 @@ on a dark background, and don't change the dark labels dark either. */
         border-color: hsla(0, 0%, 0%, 0.6);
+        .navbar-search {
+            border-left-style: none;
+            border-right-style: none;
```
forgot to delete this, this comment is a reminder to do that

timabbott commented on templates/zerver/app/notifications.html:
```diff
@@ -0,0 +1,39 @@
+            </a>
+        </div>
+    </div>
+</div>
```
Can you fix the missing newline at the end of this file?

I think we should also call this `navbar_alerts.html`; `notifications` is confusing as it suggests it's related to `notification.js` (desktop notifications).

timabbott commented on templates/zerver/app/navbar.html:
```diff
@@ -12,35 +12,37 @@
-                        <div id="search_arrows" class="pill-container input-append">
-                            <div contenteditable="true" class="input search-query input-block-level" id="search_query" type="text" placeholder="{{ _('Search') }}"
-                              autocomplete="off" aria-label="{{ _('Search') }}" title="{{ _('Search') }} (/)"> </div>
+                <div class="top-navbar-border">
```
Does this commit make any changes other than adding a div wrapper here?  Either way, you should answer that question clearly in the commit message -- that's the most important question anyone trying to review this will have since it impacts how they try to read this (large) potion of the diff.

YashRE42 commented on static/js/filter.js:
```diff
@@ -426,6 +427,147 @@ Filter.prototype = {
+        }
+    },
+
+    // get_title_expanded_list: function () {
```
Would we want to use this for eg defining the title of the Zulip tab (near the favicon)

timabbott commented:
The first three commits look great; I went ahead and merged those.

timabbott commented on static/js/filter.js:
```diff
@@ -433,6 +434,147 @@ Filter.prototype = {
+        const term_types = this.sorted_term_types();
+        if (term_types.length === 1 && ['in-home', 'in-all'].includes(term_types[0])) {return 'home';}
+        if (term_types.length === 3 && _.isEqual(term_types, ['stream', 'topic', 'search'])) {return stream_data.get_sub_by_name(this.operands('stream')[0]).name;}
+        if (term_types.length === 2 && _.isEqual(term_types, ['stream', 'topic'])) {return stream_data.get_sub_by_name(this.operands('stream')[0]).name;}
```
We generally don't use this sort of one-line return statement stylistically, and for good reason -- it makes very long lines that are hard to read.

It also seems like the 2 cases just above this have the same output and should be collapsed as an `||` in the conditional.

timabbott commented on static/js/filter.js:
```diff
@@ -433,6 +434,147 @@ Filter.prototype = {
+
+    get_icon: function () {
+        const term_types = this.sorted_term_types();
+        if (term_types.length === 1 && ['in-home', 'in-all'].includes(term_types[0])) {return 'home';}
```
Can't this go into the `switch` statement?  Seems more readable if so.

timabbott commented on static/js/filter.js:
```diff
@@ -433,6 +434,147 @@ Filter.prototype = {
+        if (term_types.length === 1 && ['in-home', 'in-all'].includes(term_types[0])) {return 'home';}
+        switch (term_types[0]) {
+        case 'stream':
+            if (stream_data.get_sub_by_name(this.operands('stream')[0]).invite_only) {
```
You end up doing this `stream_data` lookup a million times.  Would it make sense to do this lookup just at initialization time and store the result on the `Filter` object?  Seems a lot better than writing this `get_sub_by_name` thing 15 times.

timabbott commented:
I rebased this to fix the merge conflicts.

timabbott commented:
OK, I've merged most of the prep commits, including the big data commit, as the series ending with 52eaa7b5621572a6d3476e6315ef835e2117b7fc.  That was modified as follows:
* Modified by tabbott to add comments, internationalization tags on the
  strings, support streams:public, and change various title strings.

There's a several I've noticed so far that we'll want to resolve:
* This seems to fail Casper tests for a case that works in master.
* There's no whitespace in the comma-separation of users in a group PM.
* I'm not sure the variant escape behavior is exactly what we want as it feels a bit jarring; maybe we just want to always go back to the parent common narrow (if any) regardless of focus.  You can always double-escape to get to "All messages".
* What's the story with the bootstrap changes?  It's not clear to me why there's no alternative implementation to patching bootstrap.
* The `click_handlers.py` changes might be better done in an `initialize` method for `tab_bar.js`?  We're generally not adding things like this to `click_handlers.js` as it's intended for full-page elements, not individual components.
* The big TODO comment at the bottom of `tab_bar.js` -- what's your plan for that?

timabbott commented on static/templates/tab_bar.hbs:
```diff
@@ -1,19 +1,19 @@
+    <span class="sub_count" data-toggle="tooltip" title="{{sub_count}} users are subscribed to #{{title}}"><i class="fa fa-user-o"></i>{{formatted_sub_count}}</span>
+    {{/if}}
+    {{#if rendered_narrow_description}}
+    <span class="narrow_description" data-toggle="tooltip" title="{{rendered_narrow_description}}">{{{rendered_narrow_description}}}</span>
```
Use the `{{ rendered_markdown rendered_narrow_description}}` handlebars helper for this, rather than `{{{`.

timabbott commented on static/js/tab_bar.js:
```diff
@@ -1,172 +1,137 @@
+        tab_data.formatted_sub_count = get_formatted_sub_count(current_stream);
+        tab_data.stream_settings_link = "#streams/" + current_stream.stream_id + "/" + current_stream.name;
+        // legacy attribute, TODO: should be purged
+        tab_data.extra_class = 'root';
```
Can you talk about what's going on with the `extra_class` attribute?

timabbott commented on static/js/tab_bar.js:
```diff
@@ -1,172 +1,137 @@
+        const tab_bar_data = make_tab_data(filter);
+        display_tab_bar(tab_bar_data);
+        click_handlers.bind_handler_for_opening_searchbox();
+        exports.close_search_bar_and_open_narrow_description();
```
I think with early-return you could avoid duplicating `exports.close_search_bar_and_open_narrow_description();`

YashRE42 replied:
I don't think I clearly understand this.

timabbott commented:
I think we're getting close here; I think the ~10 things I mentioned here are probably fixable in this PR (maybe worth doing as new commits that can be squashed for ease of confirming the fixes), and then I'm thinking I should be able to merge the big chunk that's remaining, perhaps opening a handful of follow-up issues for things that aren't quite the way we want them but are small enough we can fix forward during a somewhat extended testing period on chat.zulip.org.

Once that's done, the next step for this component will be to migrate to the "search pills" model.

timabbott commented on static/styles/zulip.scss:
```diff
@@ -1514,9 +1513,10 @@ div.focused_table {
+                margin: 0;
+                padding: 0;
+                width: 16px;
+                height: 16px;
```
If we use the `rendered_markdown` CSS class, the need for this custom rule not be much, since the `.emoji` rendering in message bodies has shrunk.

timabbott commented on static/templates/tab_bar.hbs:
```diff
@@ -13,7 +13,7 @@
     {{/if}}
     {{#if rendered_narrow_description}}
-    <span class="narrow_description" data-toggle="tooltip" title="{{rendered_narrow_description}}">{{{rendered_narrow_description}}}</span>
+    <span class="narrow_description" data-toggle="tooltip" title="{{rendered_narrow_description}}">{{rendered_markdown rendered_narrow_description}}</span>
```
I think we want to add the `rendered_markdown` CSS class to this span; that will make this use the styling rendering of things like emoji, code blocks, etc. how they look in the message feed...

YashRE42 commented:
Okay, so I've tried to address several of the things you pointed out:

re:
* I'm not sure the variant escape behavior is exactly what we want as it feels a bit jarring; maybe we just want to always go back to the parent common narrow (if any) regardless of focus.  You can always double-escape to get to "All messages".

So, the way this currently is, is that if you have the search box selected, we execute the variant escape behavior, but if you don't have the search box selected we execute the standard `narrow.deactivate()`. I'm in support of the idea that the variant escape should be the default, but the double escape you mention is **not** a part of it. I could make it happen, though.

To be clear and explicit, this is what we're talking about implementing:
If you're in a non-common narrow: `Esc` will always return you to the parent common narrow (with the searchbox closed), regardless of whether the search bar is open or not.
If you're not in a common narrow and the searchbox is open: `Esc` will close and deselect the searchbox.
If you're not in a common narrow and the searchbox is *not* open: `Esc` will behave like `narrow.deactivate()`.

If we want to implement the above, the cleanest way would probably be as follows:
* Move the renarrowing logic to `narrow.js` as `narrow.expand_or_deactivate()`.
* Have `hotkey.process_escape_key()` call that instead of `narrow.deactivate()`.
* Leave the searchbox open/close logic in `tab_bar.js` with a call to `narrow.expand_or_deactivate()` where necessary.

re:
* What's the story with the bootstrap changes?  It's not clear to me why there's no alternative implementation to patching bootstrap.
* The big TODO comment at the bottom of `tab_bar.js` -- what's your plan for that?

So the complaints we actually wanted to address were along the lines: 
1. The typeahead should automatically be opened when the search bar is opened (both through icon / hotkey)
2. In the case that the typeahead is open, pressing `Esc` should exit both the typeahead as well as the navbar search box.

The bootstrap change achieves 2.
The test case is to make sure we're doing 1 correctly (which we're currently not).

YashRE42 commented:
Todos:
- `tab_bar.exit()` should probably reset the text in the searchbox.
- Fonts (type, size and color) are a little inconsistent between the search bar and the navbar.

follow ups:
- The lock icon to the left of stream names on recipient bars doesn't have enough right-padding (the new nav bar makes this look worse)
- Some way of indicating that we are in stream narrows (The topic is now not prominent enough in topic narrows)
- Possibly the top area should have a slight background highlight, or the separating line should be a bit bolder (or have a drop shadow)? {The new nav bar adds to our visual hierarchy problem.}
- Animated transition for opening/closing the search bar
- Behavior of compose box with respect to the navbar?
- It feels awkward that the count is between the stream name and the description.
    → maybe put it in the stream member widget and remove it from here?

timabbott commented:
OK, this is great.  I fixed a few things, did some squashing, and merged this, since I think we're better off fixing forward from here (at least in terms of using my time efficiently).  The changes were:
* The topic_name updates were buggy and removed the link (etc.).
* Removed `background-color` CSS that seemed stale from the old design.
* The rendered_description update didn't use `rendered_markdown` for rendering. (Addressed this just by removing live-update to descriptions for now).  The right fix, I'm now convinced, is to just rerender the template for live-update to things that appear in it.  (I noted that in TODOs in the code).
* Marked this as fixing a few issues that we don't intend to do further work on now that this is done.

The follow-ups on my mind are:
* Most important is fixing the live-update, I think.
* We'll start a feedback thread on chat.zulip.org for discussing further search access UI stuff.  Ultimately, though, the next big step here is rebasing/enabling @shubham-padia's pills work (i.e. killing the `searchbox_legacy` model), since that will make the search part a lot cleaner.
* We should figure out a design for handling longer stream descriptions.
* We should update the development environment test data to have more realistic initial data in `populate_db`.

I'm deploying on chat.zulip.org shortly.

timabbott commented:
And I should add, huge thanks for all your work on the marathon of getting this sort of big design change mergable.  We'll probably have a few weeks of little fixes to do, but I'm glad to have this in master.
